src/functional.rs

```diff
@@ -253,6 +256,7 @@ impl<K, V> Node<K, V> {
             right: right_child,
             value: Rc::clone(&self.value),
         }
+        .balance()
```

Comment:
This is balancing every node returned from `clone_with_children`. Not sure if that's a good place for it. It so happens that everywhere we call this we'd also want to call `balance` but it doesn't _really_ make sense as part of this method so I'm happy to manually call it everywhere.

Reply:
I actually do think it's a good place - this function just received some rando children so they may not have heights within the tolerance and needs to balance.